### PR TITLE
fix: Fix version check for mercurial

### LINKED ISSUE #363: Mercurial version is outdated!
`hg version` output:

```shell
Mercurial Distributed SCM (version 6.3.1+hg4.c44f350759c3)
(see https://mercurial-scm.org for more information)

Copyright (C) 2005-2022 Olivia Mackall and others
This is free software; see the source for copying conditions. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

But when using `:DiffviewOpen` on a hg repo, the "Mercurial version is outdated!" error will be popup.

After reading source code, I found that the `version_ok` function has a bug:

https://github.com/sindrets/diffview.nvim/blob/bff58a6ea3e081ca50049700f9848b2f84ea57be/lua/diffview/vcs/adapters/hg/init.lua#L73-L82

In my case, the `v.major` is `6` and it was greater than the `target.major`, which is `5`. But the `v.minor` is `3` that was less than the `target.minor`, which is `4`. And it causes the `version_ok` is `false`.

Seems the correct logic is:

```lua
 local version_ok = (function() 
   if v.major < target.major then 
     return false
   elseif v.major == target.major and v.minor < target.minor then 
     return false
   elseif v.major == target.major and v.minor == target.minor and v.patch < target.patch then 
     return false 
   end 
   return true 
 end)() 
```

And the `match` pattern could not extract the correct version digits:

```lua
  bs.version_string = out[1] and out[1]:match("Mercurial .*%(version (%S+)%)") or nil
```

The `bs.version_string` is `6.3.1+hg4.c44f350759c3` in my case. I think the match pattern should be `"Mercurial .*%(version (%d%.%d%.%d).*%)"`.

### PR DESCRIPTION
**Pull request by Pagliacii**
Close #363

### CODE REVIEW

**Comment by sindrets** [comment]
Thanks for the PR! This is indeed a bug. But there has to be a better way to validate this without all the redundant statements. How about something like this:

```lua
local version_ok = (function()
  if v.major ~= target.major then
    return v.major > target.major
  elseif v.minor ~= target.minor then
    return v.minor > target.minor
  elseif v.patch ~= target.patch then
    return v.patch > target.patch
  end
  return true
end)()
```

**Comment by Pagliacii** [comment]
Updated

File: lua/diffview/vcs/adapters/hg/init.lua [hunk 1]
```diff
@@ -56,7 +56,7 @@ function HgAdapter.run_bootstrap()
   end
 
   local out = utils.job(vim.tbl_flatten({ hg_cmd, "version" }))
-  bs.version_string = out[1] and out[1]:match("Mercurial .*%(version (%S+)%)") or nil
+  bs.version_string = out[1] and out[1]:match("Mercurial .*%(version (%d%.%d%.%d).*%)") or nil
```
Comment (sindrets) [inline lua/diffview/vcs/adapters/hg/init.lua:59]: This pattern won't match version numbers composed of more than a single digit.

```suggestion
  bs.version_string = out[1] and out[1]:match("Mercurial .*%(version (%d+%.%d+%.%d+)+%S*%)") or nil
```

**Review by sindrets — COMMENTED** [review]
Could you please also apply the same fix to the git adapter?

https://github.com/sindrets/diffview.nvim/blob/bff58a6ea3e081ca50049700f9848b2f84ea57be/lua/diffview/vcs/adapters/git/init.lua#L100

**Comment by Pagliacii** [comment]
Sure, I will check it at tonight.

**Comment by Pagliacii** [comment]
Updated.

Today I found that the mercurial has version likes this:

```text
Mercurial Distributed SCM (version 6.4)
(see https://mercurial-scm.org for more information)

Copyright (C) 2005-2023 Olivia Mackall and others
This is free software; see the source for copying conditions. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

So I use the `hg_version` function to extract version info from cli output. And move the **version checker** to `vcs.utils`.

**Review by sindrets — CHANGES_REQUESTED** [review]

File: lua/diffview/vcs/adapters/hg/init.lua [hunk 2]
```diff
@@ -39,6 +39,36 @@ HgAdapter.bootstrap = {
...
+    return nil
+  end
+
+  -- If patch version is not provided or is ".", set it to '0'.
+  if patch == "" or patch == "." then
+    patch = "0"
+  else
+    -- If patch version starts with ".", remove it.
+    patch = patch:sub(2)
+  end
```
[the remark this answers is not in the document]
Comment (Pagliacii) [reply lua/diffview/vcs/adapters/hg/init.lua:62]: You're right.

[the remark this answers is not in the document]
Comment (Pagliacii) [reply lua/diffview/vcs/adapters/hg/init.lua]: Updated

**Review by sindrets — APPROVED** [review]
LGTM, thank you!

**Comment by zegervdv** [comment]
@Pagliacii thanks for fixing this!
I ran into it one day, but didn't have time to debug it. The next day it was gone, so probably my version got updated in the meantime :).

Nice to see someone else is using it with mercurial.

**Pull request merged**